### PR TITLE
[SMALLFIX] Make commitBlock retry succeed

### PR DESCRIPTION
**Pull request by apc999**

### CODE REVIEW

**Comment by apc999** [comment]
Reopen of https://github.com/Alluxio/alluxio/pull/3450

File: core/server/src/main/java/alluxio/worker/block/BlockWorker.java [hunk 1]
```diff
@@ -283,7 +290,7 @@ public void commitBlock(long sessionId, long blockId)
       mBlockMasterClient.commitBlock(WorkerIdRegistry.getWorkerId(), bytesUsedOnTier,
           loc.tierAlias(), blockId, length);
     } catch (AlluxioTException | IOException | ConnectionFailedException e) {
-      throw new IOException("Failed to commit block to master.", e);
+      throw new IOException("Failed to commit block " + blockId + " to master.", e);
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/worker/block/BlockWorker.java]: Add to exception messages?

File: core/server/src/main/java/alluxio/master/block/BlockMaster.java [hunk 2]
```diff
@@ -405,6 +405,7 @@ private JournalEntry getContainerIdJournalEntry() {
    * @param length the length of the block
    * @throws NoWorkerException if the workerId is not active
    */
+  // TODO(binfan): check the logic is correct or not when commitBlock is a retry
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/block/BlockMaster.java]: Could you leave a TODO for line 466 to address the usedbytes issue later?

Comment (apc999) [inline core/server/src/main/java/alluxio/master/block/BlockMaster.java]: fixed

**Comment by apc999** [comment]
@calvinjia PTAL

**Comment by calvinjia** [comment]
LGTM

**Comment by apc999** [comment]
Jenkins, test this please

File: core/server/src/main/java/alluxio/worker/block/BlockWorker.java [hunk 3]
```diff
@@ -269,7 +270,14 @@ public void accessBlock(long sessionId, long blockId) throws BlockDoesNotExistEx
...
+    // TODO(binfan): find a better way to handle retry logic
+    try {
+      mBlockStore.commitBlock(sessionId, blockId);
+    } catch (BlockAlreadyExistsException e) {
+      LOG.debug("Block {} has been in block store, this could be a retry due to master-side RPC "
+          + "failure, therefore ignore the exception", blockId, e);
+    }
 
     // TODO(calvin): Reconsider how to do this without heavy locking.
     // Block successfully committed, update master with new block metadata
```
Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/block/BlockWorker.java]: Is updating master a second time with the same block metadata going to cause problems?

Comment (aaudiber) [inline core/server/src/main/java/alluxio/worker/block/BlockWorker.java]: Nevermind, saw your comment on the other PR

**Comment by giaosudau** [comment]
What is the version that contains this one?

**Comment by aaudiber** [comment]
@giaosudau This should be in [1.1.0](https://github.com/Alluxio/alluxio/commits/v1.1.0)

**Comment by giaosudau** [comment]
@aaudiber  but this error still happen on 1.1.0 1.1.1 1.1.2RC-2

**Comment by aaudiber** [comment]
@giaosudau What does the exception message say?

**Comment by giaosudau** [comment]
@aaudiber 
```
2016-07-20 17:56:56,238 ERROR logger.type (BlockDataServerHandler.java:handleBlockWriteRequest) - Error writing remote block : Temp blockId 166,211,878,912 is not available, because it is already committed
alluxio.exception.BlockAlreadyExistsException: Temp blockId 166,211,878,912 is not available, because it is already committed
	at alluxio.worker.block.TieredBlockStore.checkTempBlockIdAvailable(TieredBlockStore.java:393)
	at alluxio.worker.block.TieredBlockStore.createBlockMetaInternal(TieredBlockStore.java:521)
	at alluxio.worker.block.TieredBlockStore.createBlockMeta(TieredBlockStore.java:184)
	at alluxio.worker.block.BlockWorker.createBlockRemote(BlockWorker.java:336)
	at alluxio.worker.netty.BlockDataServerHandler.handleBlockWriteRequest(BlockDataServerHandler.java:145)
	at alluxio.worker.netty.DataServerHandler.channelRead0(DataServerHandler.java:71)
	at alluxio.worker.netty.DataServerHandler.channelRead0(DataServerHandler.java:40)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:244)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:294)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:831)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:322)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:254)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by aaudiber** [comment]
@giaosudau That trace seems like a different issue from this change - this change modifies the BlockWorker.commitBlock RPC, but the exception you're seeing is thrown in BlockWorker.createBlockRemote. Could you create a [JIRA](https://alluxio.atlassian.net/projects/ALLUXIO/issues/) ticket with steps to reproduce, or post a question on the [mailing list](https://groups.google.com/forum/#!forum/alluxio-users)?

**Comment by giaosudau** [comment]
@aaudiber I already did create a question on the mailing list but still haven't received anything yet.
https://groups.google.com/forum/#!topic/alluxio-users/wCPX37QS2Sg

**Pull request merged**